onedal/neighbors/neighbors.cpp

```diff
@@ -0,0 +1,286 @@
+
+    Ops ops;
+};
+
+template <typename Ops>
+struct metric2t {
+    metric2t(const Ops& ops) : ops(ops) {}
+
+    template <typename Float, typename Method>
+    auto operator()(const py::dict& params) {
+        using namespace knn;
+
+        auto metric = params["metric"].cast<std::string>();
+        ONEDAL_PARAM_DISPATCH_VALUE(metric,
+                                    "manhattan",
+                                    ops,
+                                    Float,
+                                    Method,
+                                    minkowski_distance::descriptor<Float>);
+        ONEDAL_PARAM_DISPATCH_VALUE(metric,
+                                    "minkowski",
+                                    ops,
+                                    Float,
+                                    Method,
+                                    minkowski_distance::descriptor<Float>);
+        ONEDAL_PARAM_DISPATCH_VALUE(metric,
+                                    "euclidean",
+                                    ops,
+                                    Float,
+                                    Method,
+                                    minkowski_distance::descriptor<Float>);
+        ONEDAL_PARAM_DISPATCH_VALUE(metric,
+                                    "chebyshev",
+                                    ops,
+                                    Float,
+                                    Method,
+                                    chebyshev_distance::descriptor<Float>);
+        ONEDAL_PARAM_DISPATCH_VALUE(metric,
+                                    "cosine",
+                                    ops,
+                                    Float,
+                                    Method,
+                                    cosine_distance::descriptor<Float>);
+        ONEDAL_PARAM_DISPATCH_THROW_INVALID_VALUE(metric);
+    }
+
+    Ops ops;
+};
+
+auto get_onedal_voting_mode(const py::dict& params) {
+    using namespace knn;
+
+    auto weights = params["vote_weights"].cast<std::string>();
+    if (weights == "uniform") {
+        return voting_mode::uniform;
+    }
+    else if (weights == "distance") {
+        return voting_mode::distance;
+    }
+    else
+        ONEDAL_PARAM_DISPATCH_THROW_INVALID_VALUE(weights);
+}
+
+auto get_onedal_result_options(const py::dict& params) {
+    using namespace knn;
+
+    auto result_option = params["result_option"].cast<std::string>();
+    if (result_option == "all") {
+        return result_options::responses | result_options::indices | result_options::distances;
+    }
+    else if (result_option == "indices|distances") {
+        return result_options::indices | result_options::distances;
+    }
+    else if (result_option == "responses") {
+        return result_options::responses;
+    }
+    else if (result_option == "indices") {
+        return result_options::indices;
+    }
+    else if (result_option == "distances") {
+        return result_options::distances;
+    }

```

Comment:
It can be done via let say native C++ regexes.

Reply:
https://en.cppreference.com/w/cpp/regex/regex_search

Reply:
applied, please take a look